### PR TITLE
runtime checking of eslint rules such as a default export

### PR DESCRIPTION
**Pull request by dagda1**
This PR is to add runtime checking of any eslint rules that would stop the user's build executing.

It contains 1 validator

[EslintValidator](./packages/bundler/src/validators/eslint-validator.ts).

This will call the `eslint cli engine` but only with the `@bigtest/eslint-plugin` rules.  We can't rely on the user's eslint files as rules can be disabled or removed and things like a default export are mission-critical.

For my money, any validation logic like this should utilise your bundler of choice and the thought of having this external to the bundler seems a waste of effort.

We should write plugins that utilise our bundler of choice but write them in such a way that the core logic knows nothing about the bundler.

### CODE REVIEW

**Comment by dagda1** [comment]
@cowboyd is there any feedback on this PR?

**Comment by cowboyd** [comment]
> @cowboyd is there any feedback on this PR?

Sorry, I haven't had a chance to get to this yet, but I'll make sure to take a look today

**Comment by dagda1** [comment]
> > @cowboyd is there any feedback on this PR?
> 
> Sorry, I haven't had a chance to get to this yet, but I'll make sure to take a look today

No problem, whenever you have time. 

no need to apologise.

**Comment by cowboyd** [comment]
I'm coming around to the idea of using eslint to validate the bundle, because the only way that we can reliably validate it at runtime is by evaluating it inside of context in which it will actually be used (in this case the web). The further we go, the more I expect that we won't even be able to load test trees inside Node like we do now simply because there will be modules that statically assume a browser context. I ran into this when trying to use the `localforage` library which assumes that `localStorage` is a global object and merely importing it caused a runtime exception. I had to work around it by using dynamic imports.

I suspect that we may eventually need to move the responsibility for bundling the test tree into the driver itself. 

For the moment though, it seems that the most robust way to do certain validations is with static analysis. What I'm unclear on how it all ties together is what our endgame is for validation. What I'd like to see before we add any validations to the bundle, eslint or otherwise is the design for how we account for

1. syntax errors in JS files
2. syntax errors in TS files
3. module-level evaluation errors in syntactically valid JS and TS
4. syntactically valid, module-level evaluation succeeds, but test is ill-formed
    * no default export
    * tests with same virtual path
    * tests beyond a certain depth
  
And having the simplest possible uniform representation for all of these error cases (as well as any others I may have missed). In other words, what's the optimal configuration of the GraphQL schema and the orchestrator state that supports it. We can then work backwards from there to fill in the missing pieces of what goes where.

**Comment by dagda1** [comment]
@cowboyd I am basically saying as I have done in this PR, that all validators are XXXX plugins (XXX is currently rollup).

But we write them in such a way that the core functionality knows nothing about the particular bundler.

Errors go into the atom state just like any bundler errors that we get and this is already hooked up.

I was thinking of experimenting with the typescript typechecker of forking the process.  Or would effection be a good use case for this:

```typescript
    this.service = childProcess.fork(
```

I use a webpack plugin that does this and it does cut down on build time.

You can definitely do this from a plugin of any bundler I think.

**Comment by cowboyd** [comment]
> I am basically saying as I have done in this PR, that all validators are XXXX plugins (XXX is currently rollup).
>
>But we write them in such a way that the core functionality knows nothing about the particular bundler.

What about errors that happen at a level above the bundler. For example, how do we represent it when we can't successfully load the bundle to harvest the metadata a bout it even when it bundles successfully and without errors?

Apart from that, how do you feel about making the actual eslint validator part of the eslint-plugin? It would validate the plugin architecture and decouple the core from eslint. If you're ok with that, we can tackle it in a separate PR.

**Comment by dagda1** [comment]
@cowboyd I've added the validator and the rollup plugin into the eslint-plugin package.

**Comment by dagda1** [comment]
So much better we can log now without fear of multiple identical events getting raised.

save

**Comment by jnicklas** [comment]
I've spent a lot of time this week thinking about how we can move forward with this PR and with solving #294, which this PR is trying to address.

My proposal is that we should close this PR and solve the issue that this PR is trying to address in another way. I'll try to lay out why I think this is the best way forward.

Let's look at the classification that @cowboyd posted above:

1. syntax errors in JS files
2. syntax errors in TS files (see #280)
3. module-level evaluation errors in syntactically valid JS and TS (see #592)
4. syntactically valid, module-level evaluation succeeds, but test is ill-formed
    * no default export
    * tests with same virtual path (see #294, #506)
    * tests beyond a certain depth (see #294)

I believe that (1) and (2) should be surfaced as bundling errors. I have done some experiments on integrating the `@rollup/plugin-typescript` plugin in our build pipeline, and while there are [some issues with this](https://github.com/thefrontside/bigtest/issues/280#issuecomment-704845746), I think we can make it work.

As for (3), I believe that #593 is the right way forward towards handling these types of errors. Essentially, a module level error would be handled by the agent.

This way (4) would also be handled by the *agent* which would perform a set of validations against the loaded manifest to check whether it is valid, if they cannot be verified then an error is returned. If there is no export from a test file, the agent will see this and raise an appropriate error.

This way forward, we can verify all of our assumptions without having to involve eslint.

I do think that eslint still has a useful place. The eslint plugin will be really useful for validating that there is a default export, and also other lints we might want to add in the future (a duplicate path line would be great, if that's possible), however, we won't be running eslint as part of our build pipeline. Users of bigtest will have to run eslint themselves.

In the future, we could open up for extending our rollup pipeline, so users can add whatever plugin they like. It's worth mentioning that there seems to already exist a pretty solid [eslint plugin for rollup](https://github.com/TrySound/rollup-plugin-eslint), so we don't need to maintain this ourselves.

If we move forward in this direction and we close this PR I want to nevertheless thank you @dagda1 for all of the work you put into this, it's been very helpful in fleshing out where we want to go with this. I also want to apologize for being so slow on providing my feedback here, there've been so many threads with trying to close out beta issues that I haven't had the peace of mind to really sit down and think through this problem thoroughly enough to come up with a plan.

**Comment by jnicklas** [comment]
One thing from this PR that we'll definitely want is that it surfaces the warnings generated during bundling. That is warnings generated primarily by rollup plugins. There is some code in there which stores and then displays the warnings, which is something we'll want IMO.

**Comment by dagda1** [comment]
@jnicklas IMHO relying on a default export, in general, is a bad thing.

I think the real fix is to not have to rely on this.

I personally don't think static analysis is used enough in js or typescript land and if not eslint then a babel plugin could do the same job better than any run-time check.  There are other things about the structure of the code like max tree depth that are great uses of static analysis but I can appreciate the points you raised and will definitely bow to the greater knowledge of the fuller picture.

**Comment by jnicklas** [comment]
> IMHO relying on a default export, in general, is a bad thing.

You mentioned this before, but I didn't fully understand why you were against the default export, could you elaborate on this? What would be the alternative?

> I personally don't think static analysis is used enough in js or typescript land and if not eslint then a babel plugin could do the same job better than any run-time check.

I totally agree that static analysis is good and useful, I'm a huge fan of Rust, after all.

Unfortunately while we could certainly use static analysis to give us an early warning that something is amiss, I don't think static analysis will be powerful enough to give us a definitive answer whether a test file is valid or not.

Consider the following example:

``` typescript
let t = test("foo")

if(navigator.userAgent.match(/Firefox/) {
  t = t.step("my step")
}

if(navigator.userAgent.match(/Chrome/) {
  t = t.step("my step")
}

export default t;
```

Admittedly this is a big far fetched, but note that we add steps depending on the userAgent, if we add both then we violate the "no duplicate paths" rule. This is probably perfectly fine, since we won't be adding both, but static analysis wouldn't be able to determine whether these conditions are mutually exclusive.

This example is solvable, but there are a ton of other patterns which are hard to see through with static analysis. For example, this is an evil way to get past static analysis to create an invalid test:

``` typescript
function recurse(test, number) {
  if(number === 0) {
    return test;
  } else {
    return test.child("child", (child) => recurse(child, number - 1));
  }
}

export default recurse(test("my test"), 20);
```

I think what will serve us best is a hybrid approach of having eslint rules which give an early warning in simple cases, which are easy to detect, and then having stronger validation which guards us against critical errors.

**Comment by dagda1** [comment]
That is a good example of the static analysis with only the regex literal being different but I think you could still determine that they were not the same.  Might be wrong on this one but I'm slightly interested to find out if you could :). IIRC you can parse the RegExLiteral into the actual value.  

As far as the default export goes, there are several reasons and I'll do my best to explain them.

- Having a default export has nothing to do with writing a test.  Why do I have to have a default export? (I know this answer but a new user might not).  I might be quickly having a look and then get these weird errors about default exports.

- I honestly do believe that most people in js land and ts land are agreeing that default exports are a bad thing, more so in typescript land but we are making that an absolute rule here, couple of quick examples

```ts
// source.ts
export default {}

import whatever from './source';
```

as opposed to

```ts
// source.ts
export const whatEver = {};

// typo
import { whatever } from './source'; // no exported member whatever
```

I think most are realising that named exports give better type safety.
We are going against the grain by stipulating that we need a default export.

- we are having to write code to maintain this rule and rather than trying to enforce it we should allow the user to export what they want or don't export anything which sort of feels more natural.

**Comment by cowboyd** [comment]
Does TypeScript have a way to declare an interface for a module? I mean, really a module is just a function whose input is a map of imports, and whose output is a map of exports. Seems like it should be a straightforward thing.

**Comment by dagda1** [comment]
> Does TypeScript have a way to declare an interface for a module? I mean, really a module is just a function whose input is a map of imports, and whose output is a map of exports. Seems like it should be a straightforward thing.

As ever :) there is an existing issue for [this](https://github.com/microsoft/TypeScript/issues/420)

TLDR; not currently
